### PR TITLE
Add formatter

### PR DESCRIPTION
**Pull request by yisraelU**
ready for review

### CODE REVIEW

File: modules/cli/src/runners/formatter/Formatter.scala [hunk 1]
```diff
@@ -17,52 +17,78 @@ package cli
...
-            val newPath =
-              if (noClobber) {
-                val newFile = basePath
-                  .getSegment(basePath.segmentCount - 1)
-                  .split("\\.")
-                  .mkString("_formatted.")
-                os.Path(basePath.wrapped.getParent) / s"$newFile"
-              } else basePath
+  def reformat(
+      smithyFilePath: os.Path,
```
Comment (kubukoz) [inline modules/cli/src/runners/formatter/Formatter.scala:33]:
```suggestion
      smithyWorkspacePath: os.Path,
```

File: modules/cli/src/Main.scala [hunk 2]
```diff
@@ -15,26 +15,56 @@
...
-          case opts: OpenAPIJsonSchemaOpts if !opts.isOpenapi =>
-            OpenApi.runJsonSchema(opts)
-          case opts: ProtoOpts => Proto.runFromCli(opts)
+          case OpenApiTranslate(opts) => OpenApi.runJsonSchema(opts)
+          case ProtoTranslate(opts)   => Proto.runFromCli(opts)
+          case Format(FormatOpts(files, noClobber)) =>
+            files.toList.flatMap(file => reformat(file, noClobber)).foreach {
+              case Validated.Valid(smithyFile) =>
+                println(s"Reformatted $smithyFile")
+              case Validated.Invalid(formatterError) =>
```
Comment (daddykotex) [inline modules/cli/src/Main.scala:56]: minor: you can probably extract the `FormatterError` directly with:
```scala
case Validated.Invalid(FormatterError.UnableToParse(message)) =>
  ...
case Validated.Invalid(FormatterError.InvalidModel(message)) =>
  ...
```

File: modules/cli/src/Main.scala [hunk 3]
```diff
@@ -15,26 +15,56 @@
...
+                formatterError match {
+                  case FormatterError.UnableToParse(message) =>
+                    println(
+                      s"unable to parse the Smithy file for the following reason: $message"
+                    )
+                  case FormatterError.InvalidModel(message) =>
+                    println(
+                      s"the smithy file passed in did not pass the Aws Model Validation $message"
```
Comment (daddykotex) [inline modules/cli/src/Main.scala:64]: can you tweak the message by removing `aws` here? I think `Smithy model validation: $message` would be better

File: modules/cli/src/runners/formatter/Formatter.scala [hunk 4]
```diff
@@ -0,0 +1,100 @@
...
+      } else {
+        Validated.Invalid(
+          InvalidModel(basePath.toNIO.getFileName.toString)
+        )
+      }
+
+    }
+  }
+
+  def discoverFiles(smithyFilePath: os.Path): List[(Path, String)] = {
```
Comment (daddykotex) [inline modules/cli/src/runners/formatter/Formatter.scala:72]: is it possible to reuse `FileUtils.readAll` ?

File: modules/formatter/src/ast/node.scala [hunk 5]
```diff
@@ -0,0 +1,196 @@
...
+    nodeArray.backtrack | nodeObject.backtrack | number.backtrack | node_keywords.backtrack | node_string_value.backtrack
+  )
+
+  val node_object_key: Parser[NodeObjectKey] = quoted_text.backtrack
+    .map(
+      NodeObjectKey.QuotedTextNok
+    ) | identifier.map(NodeObjectKey.IdentifierNok)
+
+  lazy val node_object_kvp: Parser[NodeObjectKeyValuePair] =
+    ((node_object_key ~ ws <* Parser.char(':')) ~ ws ~ node_value).map {
```
Comment (daddykotex) [inline modules/formatter/src/ast/node.scala:181]: I'd like it if we could move all instances of `Parser.char(/*something*/)` to the `modules/formatter/src/ast/package.scala` where others are defined

IMO, it cleans up the code quite a bit, especially when the formatter comes in and add a line in there

File: modules/formatter/src/parsers/MetadataParser.scala [hunk 6]
```diff
@@ -0,0 +1,47 @@
...
+import smithytranslate.formatter.ast.node_parser.{node_object_key, node_value}
+import smithytranslate.formatter.ast.whitespace_parser.{br, sp, sp0}
+import cats.parse.{Parser, Parser0}
+import smithytranslate.formatter.ast.MetadataStatement
+import smithytranslate.formatter.ast.MetadataSection
+
+object MetadataParser {
+  val metadata_statement: Parser[MetadataStatement] =
+    (((Parser.ignoreCase("metadata") ~ sp) *> node_object_key <* (sp0 ~ Parser
+      .char(
```
Comment (daddykotex) [inline modules/formatter/src/parsers/MetadataParser.scala:28]: like here

File: modules/formatter/src/util/string_ops.scala [hunk 7]
```diff
@@ -0,0 +1,188 @@
...
+
+    def slice(string: String, open: Char, close: Char): String = {
+      val openIndex = s.indexOf(open)
+      val closeIndex = s.indexOf(close)
+      if (openIndex < 0 || closeIndex < 0) s
+      else s.substring(openIndex + 1, closeIndex)
+
+    }
+
+    def insertAfterAll(char: Char, toInsert: String): String = {
```
Comment (daddykotex) [inline modules/formatter/src/util/string_ops.scala:60]: can be removed

File: modules/formatter/src/util/string_ops.scala [hunk 8]
```diff
@@ -0,0 +1,188 @@
...
+      ???
+    }
+  }
+
+  // todo insertion utils
+  // todo insert before character/after , single instance, all instance, first instance ,last instance, all except first , all except last
+  // index based
+  // structure detection
+  // todo isArray,isObject,isEnum
```
Comment (daddykotex) [inline modules/formatter/src/util/string_ops.scala:69]: comments still relevant?

File: modules/formatter/src/writer/IdlWriter.scala [hunk 9]
```diff
@@ -0,0 +1,31 @@
...
+
+import ast.Idl
+import ControlWriter.controlSectionWriter
+import MetadataWriter.metadataSectionWriter
+import ShapeWriter.shapeSectionWriter
+import WhiteSpaceWriter.wsWriter
+import Writer.WriterOps
+
+object IdlWriter {
+  implicit val idlWriter: Writer[Idl] = Writer.write { idl =>
```
Comment (daddykotex) [inline modules/formatter/src/writer/IdlWriter.scala:27]: can you move that instance inside the `Idl` companion object

**Review by daddykotex — COMMENTED** [review]
Great work,

I'll check it out and run it here and there to see what kind of diff it produces and report back

**Comment by daddykotex** [comment]
Ok, I just picked a random file, within this project and it failed because it was not able to validate the model. It makes sense because I have a dependency on alloy. But I'd like for the formatter to pick up automatically because it is declared in my smithy-build.json: is it out of scope for this PR ?

File: modules/formatter/src/ast/node.scala (hunk 5 above)
Comment (yisraelU) [reply modules/formatter/src/ast/node.scala:181]: all parsers are going to be under the parsers package .  there is some stuff left over , i will clean up

File: modules/cli/src/Main.scala (hunk 3 above)
Comment (daddykotex) [reply modules/cli/src/Main.scala:64]: also here, message seems to be the name of the file, which can be useful but the actual smithy error would be more interesting

**Comment by yisraelU** [comment]
> Ok, I just picked a random file, within this project and it failed because it was not able to validate the model. It makes sense because I have a dependency on alloy. But I'd like for the formatter to pick up automatically because it is declared in my smithy-build.json: is it out of scope for this PR ?

I would say yes, 
this is just an MVP to put in the correct location. Even tests are reserved for future PR's

File: modules/formatter/src/writer/IdlWriter.scala (hunk 9 above)
Comment (yisraelU) [reply modules/formatter/src/writer/IdlWriter.scala:27]: no all writers should be in the writer package , or they should all be in companion objects, due to the size of some writers ive chosen to move them all into a separate package

File: modules/cli/src/runners/formatter/Formatter.scala (hunk 4 above)
Comment (yisraelU) [reply modules/cli/src/runners/formatter/Formatter.scala:72]: possibly, i pulled out readAll for that reason, will be looking at that in the future

File: modules/cli/src/Main.scala (hunk 3 above)
Comment (yisraelU) [reply modules/cli/src/Main.scala:64]: will do

**Comment by daddykotex** [comment]
I'll write the result of my experimentations here. I ran the formatter on a few files (that had no depedencies), found a few issues. I'll mark them as TO_BE_DISCUSSED / TO_BE_FIXED. The former indicates that I don't mind a particular style, we just need to agree that it's the way to go, the latter being an problem that we need to fix IMO.


TO_BE_FIXED - double extra space before a curly brace in an operation/service declaration

Screen Shot 2022-11-21 at 15 01 08

TO_BE_DISCUSSED - alignment of closing square brackets in a multiline list

Screen Shot 2022-11-21 at 15 08 08

TO_BE_FIXED - bad handling of comments on top of struct members (for example here, `@required` ends up in the comment and renders the model invalid)

Screen Shot 2022-11-21 at 15 10 27


TO_BE_FIXED - extra new line at the end of the file

Screen Shot 2022-11-21 at 15 13 43

TO_BE_DISCUSSED - if you point the formatter to a file that does not exist, it will spit a huge exception, could we want a leaner message like: `file "path/to/file.smithy` does not exists`

Screen Shot 2022-11-21 at 15 18 05


TO_BE_FIXED - extra new line added at the end of enums

Screen Shot 2022-11-21 at 15 22 54

TO_BE_DISCUSSED - expanding empty structs

Screen Shot 2022-11-21 at 15 23 50

File: modules/cli/src/Main.scala (hunk 2 above)
Comment (yisraelU) [reply modules/cli/src/Main.scala]: i actually prefer this, i think it delineates error and success states better

**Review by daddykotex — APPROVED** [review]

**Comment by kubukoz** [comment]
Great job @yisraelU @daddykotex @lewisjkl ! 🎉

**Comment by daddykotex** [comment]
You already shared some feedback here, if you have more, make sure you share it. It aligns with most of what Michael thinks I believe: https://github.com/awslabs/smithy/pull/1531

**Comment by kubukoz** [comment]
I'll let you know if I have any more, but the "more whitespace" part should cover most of what I'd like to see changed :)

and maybe 2-space indent instead of 4.

**Pull request merged**